MCprep_addon/materials/prep.py

```diff
@@ -419,8 +423,7 @@ class MCPREP_OT_swap_texture_pack(
 
 	@classmethod
 	def poll(cls, context):
-		addon_prefs = util.get_user_preferences(context)
-		if addon_prefs.MCprep_exporter_type != "(choose)":
+		if world_tools.get_exporter(context) != world_tools.WorldExporter.Unknown:
```

Comment:
Ok so this is the reason one of the tests fail actually, and points to a behavior change.

In the currently live MCprep, someone who imports a world via the built in importer, they can still do texture swap if they manually adjust the preferences importer from (Choose) to either mineways or jmc2obj.

But in this branch, it's grayed out. 

Another use case I do sometimes: Use generate materials to create a few materials, and then later I might choose to change the texture pack for just those selected ones, even though they didn't come from an actual OBJ import. 

I know we want to ideally move away from having the global `MCprep_exporter_type` but I also want to avoid the reduction of functionality for users because we'll get a lot of people complaining that they can't use texture swap any more due to being grayed out.

Since we're still going to have the mapping json (since we need to for things like reconnect textures), I'm going to rework this so it's more of a fallback system rather than strictly the new way, since we're not removing `MCprep_exporter_type` in this release. In the future, we'll think more about how the fallback works. We could probably determine it on the fly by just checking any and all mappings in our json file for instance, basically poll to see which one is most common and then use that one (and thus, we would never really gray it out as it would always fallback to best guess for texture swap).

Reply:
Ok follow up question @StandingPadAnimations as I'm trying to do this fallback, is there a reason for this typing?

`def get_exporter(context: Context) -> Optional[WorldExporter]:`

We have some branches which return `None`, and some which return `WorldExporter.Unknown`, is there any reason to not convert the `None` response branches to `WorldExporter.Unknown`? So that anywhere we use this function, we don't have to do two checks for the kind of output (is not None and ≠ WorldExporter.Unknown)

Reply:
My local WIP which seems to be doing ok in my mind, let me know what you think. This setup makes it so it prefers the selected world if there's any data for it, falls back to scene level, and then if still nothing then if raises a runtime error (which is the test that was failing, since it was asserting we raise an error in the right situations).

```diff
diff --git a/MCprep_addon/materials/prep.py b/MCprep_addon/materials/prep.py
index 260942d..8bfc32a 100644
--- a/MCprep_addon/materials/prep.py
+++ b/MCprep_addon/materials/prep.py
@@ -425,6 +425,9 @@ class MCPREP_OT_swap_texture_pack(
 	def poll(cls, context):
 		if world_tools.get_exporter(context) != world_tools.WorldExporter.Unknown:
 			return util.is_atlas_export(context)
+		addon_prefs = util.get_user_preferences(context)
+		if addon_prefs.MCprep_exporter_type != "(choose)":
+			return util.is_atlas_export(context)
 		return False
 
 	def draw(self, context):
diff --git a/MCprep_addon/world_tools.py b/MCprep_addon/world_tools.py
index fe458f8..7831117 100644
--- a/MCprep_addon/world_tools.py
+++ b/MCprep_addon/world_tools.py
@@ -158,7 +158,7 @@ EXPORTER_MAPPING = {
 }
 
 
-def get_exporter(context: Context) -> Optional[WorldExporter]:
+def get_exporter(context: Context) -> WorldExporter:
 	"""
 	Return the exporter on the active object if it has
 	an exporter attribute.
@@ -176,7 +176,7 @@ def get_exporter(context: Context) -> Optional[WorldExporter]:
 	"""
 	obj = context.active_object
 	if not obj:
-		return None
+		return WorldExporter.Unknown
 
 	if "COMMONMCOBJ_HEADER" in obj:
 		if obj["PARENTED_EMPTY"] is not None and obj["PARENTED_EMPTY"]["exporter"] in EXPORTER_MAPPING:
@@ -195,7 +195,7 @@ def get_exporter(context: Context) -> Optional[WorldExporter]:
 		return WorldExporter.ClassicMW
 	elif prefs.MCprep_exporter_type == "jmc2obj":
 		return WorldExporter.ClassicJmc
-	return None
+	return WorldExporter.Unknown
```

Reply:
> Ok follow up question @StandingPadAnimations as I'm trying to do this fallback, is there a reason for this typing?
> 
> `def get_exporter(context: Context) -> Optional[WorldExporter]:`
> 
> We have some branches which return `None`, and some which return `WorldExporter.Unknown`, is there any reason to not convert the `None` response branches to `WorldExporter.Unknown`? So that anywhere we use this function, we don't have to do two checks for the kind of output (is not None and ≠ WorldExporter.Unknown)

`WorldExporter.Unknown` means that there is a world exporter defined in the CommonMCOBJ header, but it's not in our existing mapping of strings to enum values. `None` means no exporter can be detected at all

Reply:
To be specific, `WorldExporter.Unknown` is meant to represent any exporter that, although given in the CommonMCOBJ header, we have never tested against MCprep. This makes it so that features dependent on OBJs themselves, such as meshswap are locked, but other features aren't, such as textureswap. `None` on the other hand means no exporter is defined, so the feature set of what's possible is restricted (there can be some exceptions for legacy with the  `MCPREP_OBJ_HEADER` attribute but not the `MCPREP_OBJ_EXPORTER` attribute, but that's more of a backwards compatibility thing).

Thus, I'd rather we keep the `Optional[WorldExporter]` typing, since having that nuance between `WorldExporter.Unknown` and `None` is useful.

Reply:
Now that I think about it, looking at your diff, I think I made an error with porting textureswap. I'll fix that and push my local changes